server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java

```diff
@@ -22,138 +24,425 @@
+        return new Builder(indicesOptions);
+    }
 
-        public static EnumSet<WildcardStates> parseParameter(Object value, EnumSet<WildcardStates> defaultStates) {
-            if (value == null) {
-                return defaultStates;
+    /**
+     * Controls the way the target indices will be handled.
+     * @param allowUnavailableTargets, if false when any of the concrete targets requested does not exist, throw an error
+     */
+    public record ConcreteTargetOptions(boolean allowUnavailableTargets) implements ToXContentFragment {
+        public static final ConcreteTargetOptions ALLOW_UNAVAILABLE_TARGETS = new ConcreteTargetOptions(true);
+        public static final ConcreteTargetOptions ERROR_WHEN_UNAVAILABLE_TARGETS = new ConcreteTargetOptions(false);
+
+        public static ConcreteTargetOptions fromParameter(Object ignoreUnavailableString, ConcreteTargetOptions defaultOption) {
+            if (ignoreUnavailableString == null && defaultOption != null) {
+                return defaultOption;
             }
+            return nodeBooleanValue(ignoreUnavailableString, "ignore_unavailable")
+                ? ALLOW_UNAVAILABLE_TARGETS
+                : ERROR_WHEN_UNAVAILABLE_TARGETS;
+        }
 
-            EnumSet<WildcardStates> states = EnumSet.noneOf(WildcardStates.class);
-            String[] wildcards = nodeStringArrayValue(value);
-            // TODO why do we let patterns like "none,all" or "open,none,closed" get used. The location of 'none' in the array changes the
-            // meaning of the resulting value
-            for (String wildcard : wildcards) {
-                updateSetForValue(states, wildcard);
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            return builder.field("ignore_unavailable", allowUnavailableTargets);
+        }
+    }
+
+    /**
+     * Controls the way the wildcard expressions will be resolved.
+     * @param matchOpen, open indices will be matched
+     * @param matchClosed, closed indices will be matched
+     * @param includeHidden, hidden indices will be included in the result. This is a post filter, it requires matchOpen or matchClosed
+     *                      to have an effect.
+     * @param resolveAliases, aliases will be included in the result, if false we treat them like they do not exist
+     * @param allowEmptyExpressions, when an expression does not result in any indices, if false it throws an error if true it treats it as
+     *                               an empty result
+     */
+    public record WildcardOptions(
+        boolean matchOpen,
+        boolean matchClosed,
+        boolean includeHidden,
+        boolean resolveAliases,
+        boolean allowEmptyExpressions
+    ) implements ToXContentFragment {
+
+        public static final WildcardOptions DEFAULT = WildcardOptions.newBuilder().build();
```

Comment:
With the other comment about enumeration of the defaults, I think I'd prefer to enumerate all of the defaults here, so a developer can see them, and then make the `WildcardOptions.Builder` reference them, rather than the other way around